ffi/capi/src/locale.rs

```diff
@@ -82,56 +62,34 @@ pub extern "C" fn icu4x_locale_basename(
 
 #[no_mangle]
 /// Write a string representation of the calendar extension to `write`
-pub extern "C" fn icu4x_locale_calendar(
-    locale: &ICU4XLocale,
-    write: &mut ICU4XWriteable,
-) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("ca"), locale, write)
-}
-
-#[no_mangle]
-/// Write a string representation of the case first extension to `write`
-pub extern "C" fn icu4x_locale_case_first(
-    locale: &ICU4XLocale,
-    write: &mut ICU4XWriteable,
-) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("kf"), locale, write)
-}
-
-#[no_mangle]
-/// Write a string representation of the collation extension to `write`
-pub extern "C" fn icu4x_locale_collation(
-    locale: &ICU4XLocale,
-    write: &mut ICU4XWriteable,
-) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("co"), locale, write)
-}
-
-#[no_mangle]
-/// Write a string representation of the hour cycle extension to `write`
-pub extern "C" fn icu4x_locale_hour_cycle(
-    locale: &ICU4XLocale,
-    write: &mut ICU4XWriteable,
-) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("hc"), locale, write)
-}
-
-#[no_mangle]
-/// Write a string representation of the numeric extension to `write`
-pub extern "C" fn icu4x_locale_numeric(
-    locale: &ICU4XLocale,
-    write: &mut ICU4XWriteable,
-) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("kn"), locale, write)
-}
-
-#[no_mangle]
-/// Write a string representation of the number system extension to `write`
-pub extern "C" fn icu4x_locale_numbering_system(
+///
+/// # Safety
+/// `value` and `len` should point to a valid ASCII string of length `len`.
+///
+/// It does not need to be be null terminated, and `len` should not include a null
+/// terminator (this will just cause the function to panic, and is not a safety requirement).
+pub unsafe extern "C" fn icu4x_locale_get_extension(
     locale: &ICU4XLocale,
+    value: *const u8,
+    len: usize,
     write: &mut ICU4XWriteable,
 ) -> ICU4XLocaleResult {
-    write_extension!(tinystr4!("nu"), locale, write)
+    let bytes = slice::from_raw_parts(value, len);
+    if let Ok(key) = Key::from_bytes(bytes) {
+        if let Some(value) = locale.extensions.unicode.keywords.get(&key) {
```

Comment:
suggestion: This should still be a convenience method on Locale IMO that gets thinly wrapped here, the FFI layer shouldn't ever have APIs the Rust layer doesn't.